samples/net/sockets/http_server/src/certs/gen_ca_cert.sh

```diff
@@ -0,0 +1,17 @@
+# Copyright (c) 2024, Witekio
```

Comment:
I was thinking about this, I wonder if it's possible to generate the certificates using the python `cryptography` module, which is already in the python requirements for `imgtool`, completely removing the need for `openssl`.

Probably a task for the future though, I'm not sure where I would start with this...

Reply:
> Probably a task for the future though,

Yes, definitely a future work.

Reply:
How about leaving the cert files off, maybe have the sample auto generate them if they are not find, and add a big warning that this only works on Linux out of the box? I understand that these files were there before but I think there's value in getting rid of them, I'd hate to see these landing a product because "oh I'll generate proper certs later".

Reply:
Regarding this getting into an end product, I think it's much lower risk than e.g. the default signing keys for mcuboot. Perhaps it could be handled using a similar build-time warning to the mcuboot one:
```
  WARNING: Using default MCUboot signing key file, this file is for debug use
  only and is not secure!
```

Although now that I think about it mcuboot is embedding a key from a PEM file in the software, so perhaps we already have a way in-tree to do the PEM --> C array conversion and we could do away with the `.der` files and convert from `.pem` at build time. This would avoid the compliance issue but obviously we'd still have the same default certificates, just in a different format.

There's currently 24 `.der` files in tree, mostly across networking samples, so maybe a common approach across all of these would be useful.

Having said all that, autogenerating the certificates at build time if they don't already exist should be quite straightforward, I'm happy with that approach if it's the one agreed on.

Reply:
The built time configuration would be nice, but how do we do that with CI, are all the needed tools available during build, and how much longer the build would take.
The problem with PEM format is that then we need to embed PEM parser from mbedtls and it takes ram and rom.

In the products the certs should be handled differently and not embed them into the application. We have the cert storage that should be used to store them.

I have been pondering this idea that is it possible to have just one set of certs for all the network samples. Not sure if it possible, but then we could then avoid putting these certs into every sample that needs them.

Reply:
> The built time configuration would be nice, but how do we do that with CI, are all the needed tools available during build, and how much longer the build would take.

Is openssl not on the CI image, sure it can be added. As for build time, is that a problem? Ran the script below seems like they are done ~instantly.

Having a common set of certs for all network samples is certainly a good idea and should be done regardless, and so is having the pem to der conversion done at build time.

Reply:
The CI docker container already contains `openssl` and the scripts seem to run fine, the scripts themselves take only a few 10's of milliseconds to run. Although I don't think the HTTP server sample is currently built with HTTPS enabled in CI anyway, so the certificate generation would not be needed.

> The problem with PEM format is that then we need to embed PEM parser from mbedtls and it takes ram and rom.

I was thinking we could do the conversion at build time so there is no change to the application itself. However after checking how `mcuboot` does the conversion from `CONFIG_BOOT_SIGNATURE_KEY_FILE` --> `autogen-pubkey.c` it uses `imgtool`, which can only extract the public key from a key file rather than being a more generic conversion from PEM format which could handle certificates as well as just keys. So we can't use the same mechanism directly.

Reply:
you can probably use `openssl x509` in CI? from a CMake target

Reply:
`openssl` from a CMake target works, however currently we only have the `openssl` dependency if the user wants to generate their own certificates.

Replacing the DER default certs with PEM would make `openssl` a requirement for building the sample with HTTPS at all, even using the default certificates (not necessarily against this, but it means that we can no longer build some things on Windows that used to work)

Reply:
Personally, I'd just stick to pre-generated certs in the sample. It's really hard for me to imagine someone using a publicly available private key from a demo app in their end products.

Reply:
When you say stick to the pre-generated certs, do you mean to keep the PR as-is, or remove the certificate generation option entirely?

The main reason I added the ability to generate your own certificates is not for the device side of things, but so that the user doesn't have to add the provided "demo" CA certificate to the list of trusted CA certificates in their browser. This would definitely be a security risk - in theory I could sign whatever I like using the associated private key and the user's browser would trust it (in practice I've deleted the private key already, but nobody should have to trust that)

Reply:
Keep as is, I don't mind having an **option** to generate own certs.

Reply:
Alright sounds good, thanks for the information dump.

@rlubos @jukkar I think we can go with this now, you need to ask @carlescufi or someone else to merge, but do you think it would be an option to followup and at least have a single set of certificates for, let's say, all samples in `net/sockets`?

```
./net/sockets/echo_server/src/server_privkey.der
./net/sockets/echo_server/src/server.der
./net/sockets/echo_server/src/echo-apps-cert.der
./net/sockets/echo_server/src/ca.der
./net/sockets/echo_server/src/echo-apps-key.der
./net/sockets/dumb_http_server_mt/src/mt-http-server-key.der
./net/sockets/dumb_http_server_mt/src/mt-http-server-cert.der
./net/sockets/http_server/src/server_privkey.der
./net/sockets/http_server/src/https-server-cert.der
./net/sockets/http_server/src/server.der
./net/sockets/http_server/src/https-server-key.der
./net/sockets/http_server/src/ca.der
./net/sockets/http_client/src/https-cert.der
./net/sockets/echo_client/src/echo-apps-cert.der
```

At least we would minimize the number of these kicking around in the repo.

Reply:
Why can't we provide a PowerShell script for Windows users? AFAIK this is self-contained and doesn't require e.g. openssl
https://learn.microsoft.com/en-us/powershell/module/pki/new-selfsignedcertificate